File: sql-insight-core/src/main/java/org/gongxuanzhang/sql/insight/core/object/TableExt.java
/*
 * Copyright 2023 java-mysql  and the original author or authors <gongxuanzhangmelt@gmail.com>.
 *
 * Licensed under the GNU Affero General Public License v3.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://github.com/implement-study/sql-insight/blob/main/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gongxuanzhang.sql.insight.core.object;

import lombok.Data;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author gongxuanzhangmelt@gmail.com
 **/
@Data
public class TableExt {

    final Map<String, Column> columnMap = new HashMap<>();

    final Map<String, Integer> columnIndex = new HashMap<>();

    /**
     * not null column index list
     **/
    final List<Integer> notNullIndex = new ArrayList<>();

    final List<Integer> variableIndex = new ArrayList<>();

    int autoColIndex = -1;

    int primaryKeyIndex = -1;

    int nullableColCount;




}

File: sql-insight-core/src/main/kotlin/tech/insight/core/bean/desc/TableExt.kt
package tech.insight.core.bean.desc

import tech.insight.core.bean.Column
import tech.insight.core.bean.SQLBean
import tech.insight.core.bean.Table

/**
 *
 * table extension info
 *
 * @author gongxuanzhangmelt@gmail.com
 */
class TableExt(val table: Table) : SQLBean {

    val columnMap: MutableMap<String, Column> = HashMap()
    val columnIndex: MutableMap<String, Int> = HashMap()

    /**
     * not null column index list
     */
    val notNullIndex: MutableList<Int> = ArrayList()
    val variableIndex: MutableList<Int> = ArrayList()
    var autoColIndex = -1
    var primaryKeyIndex = -1
    var nullableColCount = 0
    var primaryKeyName: String? = null


    init {
        table.columnList.forEachIndexed { index, col ->
            columnMap[col.name] = col
            columnIndex[col.name] = index
            if (col.primaryKey) {
                this.primaryKeyIndex = index
                this.primaryKeyName = col.name
            }
            if (col.autoIncrement) {
                require(autoColIndex == -1) { "auto increment column can have at most one" }
                autoColIndex = index
            }
            if (col.notNull) {
                notNullIndex.add(index)
            } else {
                this.nullableColCount++
            }
            if (col.variable) {
                this.variableIndex.add(index)
            }
        }
    }

    override fun parent(): SQLBean {
        return table
    }
}
